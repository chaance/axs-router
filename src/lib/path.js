import { pick, segmentize, startsWith } from './utils';
import { PARAM_REGEX } from './constants';

function addQuery(pathname, query) {
  return pathname + (query ? `?${query}` : '');
}

////////////////////////////////////////////////////////////////////////////////
// insertParams(path, params)
function insertParams(path, params) {
  let segments = segmentize(path);
  return (
    '/' +
    segments
      .map(segment => {
        let match = PARAM_REGEX.exec(segment);
        return match ? params[match[1]] : segment;
      })
      .join('/')
  );
}

////////////////////////////////////////////////////////////////////////////////
// match(path, uri) - Matches just one path to a uri, also lol
function match(path, uri) {
  return pick([{ path }], uri);
}

////////////////////////////////////////////////////////////////////////////////
// resolve(to, basepath)
//
// Resolves URIs as though every path is a directory, no files.  Relative URIs
// in the browser can feel awkward because not only can you be "in a directory"
// you can be "at a file", too. For example
//
//     browserSpecResolve('foo', '/bar/') => /bar/foo
//     browserSpecResolve('foo', '/bar') => /foo
//
// But on the command line of a file system, it's not as complicated, you can't
// `cd` from a file, only directories.  This way, links have to know less about
// their current path. To go deeper you can do this:
//
//     <Link to="deeper"/>
//     // instead of
//     <Link to=`{${props.uri}/deeper}`/>
//
// Just like `cd`, if you want to go deeper from the command line, you do this:
//
//     cd deeper
//     # not
//     cd $(pwd)/deeper
//
// By treating every path as a directory, linking to relative paths should
// require less contextual information and (fingers crossed) be more intuitive.
function resolve(to, base) {
  // /foo/bar, /baz/qux => /foo/bar
  if (startsWith(to, '/')) {
    return to;
  }

  let [toPathname, toQuery] = to.split('?');
  let [basePathname] = base.split('?');

  let toSegments = segmentize(toPathname);
  let baseSegments = segmentize(basePathname);

  // ?a=b, /users?b=c => /users?a=b
  if (toSegments[0] === '') {
    return addQuery(basePathname, toQuery);
  }

  // profile, /users/789 => /users/789/profile
  if (!startsWith(toSegments[0], '.')) {
    let pathname = baseSegments.concat(toSegments).join('/');
    return addQuery((basePathname === '/' ? '' : '/') + pathname, toQuery);
  }

  // ./         /users/123  =>  /users/123
  // ../        /users/123  =>  /users
  // ../..      /users/123  =>  /
  // ../../one  /a/b/c/d    =>  /a/b/one
  // .././one   /a/b/c/d    =>  /a/b/c/one
  let allSegments = baseSegments.concat(toSegments);
  let segments = [];
  for (let i = 0, l = allSegments.length; i < l; i++) {
    let segment = allSegments[i];
    if (segment === '..') segments.pop();
    else if (segment !== '.') segments.push(segment);
  }

  return addQuery('/' + segments.join('/'), toQuery);
}

export { addQuery, insertParams, match, resolve };
