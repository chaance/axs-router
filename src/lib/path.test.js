import { match, resolve, insertParams } from "./path";

describe("match", () => {
  test("works", () => {
    expect(match("/foo/:bar", "/foo/hello")).toMatchSnapshot();
  });
});

describe("resolve", () => {
  it("works with a / base", () => {
    expect(resolve("contacts/ryan", "/")).toEqual("/contacts/ryan");
  });

  test("a bunch of resolves", () => {
    expect(resolve("/somewhere/else", "/users/123")).toEqual("/somewhere/else");
    expect(resolve("settings", "/users/123")).toEqual("/users/123/settings");
    expect(resolve("../../one/../two/.././three", "/a/b/c/d/e/f/g")).toEqual(
      "/a/b/c/d/e/three"
    );
    expect(resolve("./", "/users/123")).toEqual("/users/123");
    expect(resolve(".", "/users/123")).toEqual("/users/123");
    expect(resolve("../", "/users/123")).toEqual("/users");
    expect(resolve("../..", "/users/123")).toEqual("/");
    expect(resolve("../.././3", "/u/1/g/4")).toEqual("/u/1/3");
    expect(resolve("../.././s?beef=boof", "/u/1/g/4")).toEqual(
      "/u/1/s?beef=boof"
    );
    expect(resolve("../.././3", "/u/1/g/4?beef=boof")).toEqual("/u/1/3");
    expect(resolve("stinky/barf", "/u/1/g/4")).toEqual("/u/1/g/4/stinky/barf");
    expect(resolve("?some=query", "/users/123?some=thing")).toEqual(
      "/users/123?some=query"
    );
    expect(resolve("/groups?some=query", "/users?some=thing")).toEqual(
      "/groups?some=query"
    );
  });
});

describe("insertParams", () => {
  it("works", () => {
    expect(
      insertParams("/users/:userId/groups/:groupId", {
        userId: "2",
        groupId: "4"
      })
    ).toEqual("/users/2/groups/4");
  });
});
