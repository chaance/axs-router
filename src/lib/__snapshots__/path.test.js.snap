// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`match works 1`] = `
Object {
  "params": Object {
    "bar": "hello",
  },
  "route": Object {
    "path": "/foo/:bar",
  },
  "uri": "/foo/hello",
}
`;
